fpnge.cc

```diff
@@ -543,40 +539,55 @@ FORCE_INLINE MIVEC PredictVec(const unsigned char *current_buf,
   }
 }
 
+alignas(SIMD_WIDTH) const int32_t _kMaskVec[] = {-1, -1, -1, -1,
+#if SIMD_WIDTH == 32
+                                                 -1, -1, -1, -1, 0, 0, 0, 0,
+#endif
+                                                 0,  0,  0,  0};
+const uint8_t *kMaskVec = (const uint8_t *)_kMaskVec + SIMD_WIDTH;
+
 template <size_t predictor, typename CB, typename CB_ADL, typename CB_RLE>
 static FORCE_INLINE void
-ProcessRow(size_t bytes_per_line_buf, const unsigned char *mask,
-           const unsigned char *current_row_buf, const unsigned char *top_buf,
-           const unsigned char *left_buf, const unsigned char *topleft_buf,
-           CB &&cb, CB_ADL &&cb_adl, CB_RLE &&cb_rle) {
+ProcessRow(size_t bytes_per_line, const unsigned char *current_row_buf,
+           const unsigned char *top_buf, const unsigned char *left_buf,
+           const unsigned char *topleft_buf, CB &&cb, CB_ADL &&cb_adl,
+           CB_RLE &&cb_rle) {
   size_t run = 0;
-  for (size_t i = 0; i + SIMD_WIDTH <= bytes_per_line_buf; i += SIMD_WIDTH) {
+  size_t i = 0;
+  for (; i + SIMD_WIDTH <= bytes_per_line; i += SIMD_WIDTH) {
     auto pdata = PredictVec<predictor>(current_row_buf + i, top_buf + i,
                                        left_buf + i, topleft_buf + i);
-    auto maskv = MMSI(load)((MIVEC *)(mask + i));
-
-    size_t bytes_per_vec = _mm_popcnt_u32(MM(movemask_epi8)(maskv));
-
-    auto pdatais0 = MM(cmpeq_epi8)(pdata, MMSI(setzero)());
-    auto isnot0 = MMSI(andnot)(pdatais0, maskv);
-
-    unsigned next0run =
-#ifdef __AVX2__
-        _tzcnt_u32(MM(movemask_epi8)(isnot0));
-#else
-        BSF32((1UL << SIMD_WIDTH) | MM(movemask_epi8)(isnot0));
-#endif
+    unsigned pdatais0 =
+        MM(movemask_epi8)(MM(cmpeq_epi8)(pdata, MMSI(setzero)()));
+    if (pdatais0 == SIMD_MASK) {
+      run += SIMD_WIDTH;
+    } else {
+      if (run != 0) {
+        cb_rle(run);
+      }
+      run = 0;
+      cb(pdata, SIMD_WIDTH);
+    }
+    cb_adl(SIMD_WIDTH, pdata, SIMD_WIDTH, i);
+  }
+  size_t bytes_remaining = bytes_per_line ^ i;
```

Comment:
Any reason why this is not a - instead of a ^?

Reply:
Both work in this case. In such a case, XOR is at least as good as subtract, if not better, due to:

* some processors have more ALUs that can do bitwise operations than arithmetic - never the other way around (only possible exception on x86 is the `LEA` instruction, which can be used to save a move, but not relevant here)
* XOR is commutative, so the compiler has more freedom with register ordering (particularly since x86 is a 2-operand ISA)

Might not matter much here, but either works.

Reply:
Can we either use - or add a comment?